### PR TITLE
Enable the use of batch api using multiple speakers with Bark

### LINKED ISSUE #26921: BarkModel: support multiple history prompt for batch api
### Feature request

Add multiple history prompt support for batch api

### Motivation

Batch API speeds up the audio generation process a lot. Unfortunately we can't use it to generate long sequences.

### Your contribution

I may be able to look at it soon

### PR DESCRIPTION
**Pull request by MidAtBest**
# What does this PR do?

This PR aims to enable the use of the batch API when using multiple speaker prompts for Bark.

Currently, the batch API is available when using multiple text inputs but is limited to a single speaker. The objective of this PR is to enable this when multiple speaker prompts are being used. This is done keeping in mind that we want to minimise the amount of loops when processing the data while keeping the same audio quality.

cc @ylacombe  and @Selectorrr , could you take a look ? Thanks!

Fixes #26921

PS: Currently I have doubts on how to solve the issue in the `BarkCoarseModel` especially these 2 lines:

```
input_coarse = semantic_output[:, np.max([0, semantic_idx - max_semantic_history]) :]
input_coarse = input_coarse[:, :max_coarse_input_length]
```

My concern is that with the approach implemented in this PR, if we need to pad after the `preprocess_histories` method because the samples in the batch for `x_coarse` are not the same length then we might run into a scenario where some padding tokens would be used in `input_coarse` instead of "regular tokens" if the sample was processed outside of a batch.

### CODE REVIEW

**Review by ylacombe — COMMENTED** [review]
Hi @MidAtBest, thanks for taking care of this, this is a great PR! 
The logic is quite difficult so congrats on handling it so nicely !

Two general comments: 
1. you should add test for bark modeling code as well !
2. it's best to stay with `voice_preset` instead of `voice_presets`.

I'm responding to your question below, I hope that it helps!

Thanks again!

---------------

In terms of your question, I believe that we should refactor the logic of the coarse generation quite a bit to make it work. 

The key to that is IMO in the following lines:

These ones creates `semantic_output`:
https://github.com/huggingface/transformers/blob/87ae2a4632de9c3090272d5cd37db86c0e03a0a9/src/transformers/models/bark/modeling_bark.py#L1175-L1185

Those 3 lines are located in the generation loop and creates `input_coarse`:
https://github.com/huggingface/transformers/blob/87ae2a4632de9c3090272d5cd37db86c0e03a0a9/src/transformers/models/bark/modeling_bark.py#L1196-L1204

As you can see, the only thing that changes here is `semantic_idx`, the rest is quite static.

So in my opinion, you keep track of the length of each `semantic_history` and merge each `semantic_history` with the corresponding `semantic_output`.

Once arriving in the generation loop, you can do the 2 first lines for each sample (taking into account than each sample has its own `semantic_idx`) then use `torch.rnn.pad_sequence` to pad it to the right length.

**Comment by MidAtBest** [comment]
> Hi @MidAtBest, thanks for taking care of this, this is a great PR! The logic is quite difficult so congrats on handling it so nicely !
> 
> Two general comments:
> 
> 1. you should add test for bark modeling code as well !
> 2. it's best to stay with `voice_preset` instead of `voice_presets`.
> 
> I'm responding to your question below, I hope that it helps!
> 
> Thanks again!
> 
> In terms of your question, I believe that we should refactor the logic of the coarse generation quite a bit to make it work.
> 
> The key to that is IMO in the following lines:
> 
> These ones creates `semantic_output`:
> 
> https://github.com/huggingface/transformers/blob/87ae2a4632de9c3090272d5cd37db86c0e03a0a9/src/transformers/models/bark/modeling_bark.py#L1175-L1185
> 
> Those 3 lines are located in the generation loop and creates `input_coarse`:
> 
> https://github.com/huggingface/transformers/blob/87ae2a4632de9c3090272d5cd37db86c0e03a0a9/src/transformers/models/bark/modeling_bark.py#L1196-L1204
> 
> As you can see, the only thing that changes here is `semantic_idx`, the rest is quite static.
> 
> So in my opinion, you keep track of the length of each `semantic_history` and merge each `semantic_history` with the corresponding `semantic_output`.
> 
> Once arriving in the generation loop, you can do the 2 first lines for each sample (taking into account than each sample has its own `semantic_idx`) then use `torch.rnn.pad_sequence` to pad it to the right length.

Thanks a lot for the reply, yes it took a bit of time for me to understand both the model and the shape manipulations going on. 
Regarding the tests and other changes I had omitted them as I was a bit puzzled regarding the question I mentioned in the body of the PR. 
Setting the PR as WIP and will try to have something by Wednesday evening.

File: src/transformers/models/bark/modeling_bark.py [hunk 1]
```diff
@@ -961,20 +961,28 @@ def generate(
...
+                        sp[-max_input_semantic_length:],
+                        (0, max_input_semantic_length - len(sp[-max_input_semantic_length:])),
+                        value=semantic_generation_config.semantic_pad_token,
+                        mode="constant",
+                    ).unsqueeze(0)
+                    for sp in history_prompt["semantic_prompt"]
+                ],
+                dim=0,
+            ).to(self.device)
+
```
[the remark this answers is not in the document]
Comment (MidAtBest) [reply src/transformers/models/bark/modeling_bark.py]: After spending some time looking into this into more details, there is still the issue driven by the fact that the `semantic_output` is not the same length if I process 2 samples simultaneously instead of processing it separately.

Hence, the following:

```
voice_preset = "v2/fr_speaker_0"
inputs = processor("Je m'appelle Bob.", voice_preset=voice_preset)

model = BarkModel.from_pretrained("suno/bark")
audio_array = model.generate(**inputs, do_sample=False)
```

will lead to `semantic_output` smaller Than the tensor associated to the same prompt and voice preset in the following case:

```
processor = AutoProcessor.from_pretrained("suno/bark")

voice_preset = ["v2/fr_speaker_0", "v2/en_speaker_0"]

inputs = processor(["Je m'appelle Bob.", "Yo, bro, my name is bobby boy"], voice_preset=voice_preset)

model = BarkModel.from_pretrained("suno/bark")
audio_array = model.generate(**inputs, do_sample=False, temperature=1)
```

The outputs overall make sense and seem to match in both cases but this difference is still unwanted. 

A solution would be to remove the end padding for the `semantic_output` and do the processing as you mentioned and as I have currently implemented for `BarkCoarseModel`. And potentially do the same for the `coarse_output` in the case of `BarkFineModel` if necessary.

File: src/transformers/models/bark/modeling_bark.py [hunk 2]
```diff
@@ -1172,14 +1163,61 @@ def generate(
...
+            if len(x_semantic_history) == 1:
+                x_semantic_history = torch.repeat_interleave(x_semantic_history, batch_size, dim=0).to(self.device)
+                x_coarse = torch.repeat_interleave(x_coarse, batch_size, dim=0).to(self.device)
+
+        else:
+            x_semantic_history = torch.tensor([[]] * batch_size, dtype=torch.int).to(self.device)
+            x_coarse = torch.tensor([[]] * batch_size, dtype=torch.int).to(self.device)
+
```
[the remark this answers is not in the document]
Comment (MidAtBest) [reply src/transformers/models/bark/modeling_bark.py]: Good point, it's done :)

File: src/transformers/models/bark/modeling_bark.py [hunk 3]
```diff
@@ -1172,14 +1163,61 @@ def generate(
...
+                    nn.functional.pad(
+                        x_ch,
+                        (max_len_ch - len(x_ch), 0),
+                        value=coarse_generation_config.coarse_semantic_pad_token,
+                        mode="constant",
+                    )
+                    for x_ch in x_coarse
+                ],
+                dim=0,
+            ).to(self.device)
```
[the remark this answers is not in the document]
Comment (MidAtBest) [reply src/transformers/models/bark/modeling_bark.py]: Removed it for `x_semantic_history` as we are processing the samples separately, and slightly adapted it  for `x_coarse` as we want to do left padding as later on we deal with the following  ` x_coarse[:, -max_coarse_history:]`.

I am not certain that this is the best option though.

File: src/transformers/models/bark/processing_bark.py [hunk 4]
```diff
@@ -219,7 +220,7 @@ def _validate_voice_preset_dict(self, voice_preset: Optional[dict] = None):
     def __call__(
         self,
         text=None,
-        voice_preset=None,
+        voice_presets=None,
```
[the remark this answers is not in the document]
Comment (MidAtBest) [reply src/transformers/models/bark/processing_bark.py]: Understood, done.

File: src/transformers/models/bark/processing_bark.py [hunk 5]
```diff
@@ -238,7 +239,7 @@ def __call__(
                 The sequence or batch of sequences to be encoded. Each sequence can be a string or a list of strings
                 (pretokenized string). If the sequences are provided as list of strings (pretokenized), you must set
                 `is_split_into_words=True` (to lift the ambiguity with a batch of sequences).
-            voice_preset (`str`, `Dict[np.ndarray]`):
+            voice_preset (`str`, `Dict[np.ndarray]`, List[Dict[np.ndarray]]):
```
[the remark this answers is not in the document]
Comment (MidAtBest) [reply src/transformers/models/bark/processing_bark.py]: Yes, adapted the docstring!

File: src/transformers/models/bark/processing_bark.py [hunk 6]
```diff
@@ -238,7 +239,7 @@ def __call__(
...
                 The voice preset, i.e the speaker embeddings. It can either be a valid voice_preset name, e.g
                 `"en_speaker_1"`, or directly a dictionnary of `np.ndarray` embeddings for each submodel of `Bark`. Or
                 it can be a valid file name of a local `.npz` single voice preset.
```
[the remark this answers is not in the document]
Comment (MidAtBest) [reply src/transformers/models/bark/processing_bark.py]: Done

File: src/transformers/models/bark/processing_bark.py [hunk 7]
```diff
@@ -252,23 +253,42 @@ def __call__(
             Tuple([`BatchEncoding`], [`BatchFeature`]): A tuple composed of a [`BatchEncoding`], i.e the output of the
             `tokenizer` and a [`BatchFeature`], i.e the voice preset with the right tensors type.
```
[the remark this answers is not in the document]
Comment (MidAtBest) [reply src/transformers/models/bark/processing_bark.py]: Changed it I am now using [NestedTensors](https://pytorch.org/tutorials/prototype/nestedtensor.html) to initialise the `BatchFeature` object.

File: tests/models/bark/test_processor_bark.py [hunk 8]
```diff
@@ -82,30 +82,88 @@ def test_speaker_embeddings(self):
...
+        inputs = processor(text=self.input_string, voice_presets=tmpfilenames)
+        processed_voice_presets = inputs["history_prompt"]
+
+        for i, voice_preset in enumerate(voice_presets):
+            for key in voice_preset:
+                self.assertListEqual(
+                    voice_preset[key].tolist(), processed_voice_presets.get(key, np.array([]))[i].tolist()
+                )
+        # test loading voice preset from the hub
+        inputs = processor(text=self.input_string, voice_presets=self.voice_preset)
```
[the remark this answers is not in the document]
Comment (MidAtBest) [reply tests/models/bark/test_processor_bark.py]: Done and additionally added a test that verifies if it breaks in the case where the number of `voice_preset` is different than 1 and `batch_size`.

File: src/transformers/models/bark/modeling_bark.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ylacombe) [reply src/transformers/models/bark/modeling_bark.py]: Hey @MidAtBest, not sure to follow what's the issue here, can you expand on this ?

**Comment by ArthurZucker** [comment]
cc @ylacombe would be nice if you can review here!

**Comment by amyeroberts** [comment]
Gentle ping @ylacombe

**Comment by amyeroberts** [comment]
Another ping @ylacombe

**Comment by amyeroberts** [comment]
Another ping @ylacombe

**Comment by qgallouedec** [comment]
closing see https://github.com/huggingface/transformers/issues/26921#issuecomment-5704520295
